src/headset/oculus.c

```diff
@@ -0,0 +1,531 @@
+#include <OVR_CAPI.h>
+#include <OVR_CAPI_GL.h>
+
+typedef struct {
+  bool isInitialized;
+  bool isRendering;
+  bool isMirrored;
+  bool hmdPresent;
+  headsetRenderCallback renderCallback;
+  vec_controller_t controllers;
+  ovrSession session;
+  ovrGraphicsLuid luid;
+  ovrTrackingState ts;
+  ovrInputState is;
+  int frameIndex;
+  float clipNear;
+  float clipFar;
+  int lastButtonState;
+  struct {
+    ovrSizei size;
+    GLuint fboId;
+    GLuint depthId;
+    ovrTextureSwapChain chain;
+  } eyeTextures[2];
+  ovrMirrorTexture mirrorTexture;
+  GLuint mirrorFBO;
+} HeadsetState;
+
+static HeadsetState state;
+
+void lovrHeadsetInit() {
+  vec_init(&state.controllers);
+  ovrResult result = ovr_Initialize(NULL);
+  if (OVR_FAILURE(result))
+    return;
+
+  result = ovr_Create(&state.session, &state.luid);
+  if (OVR_FAILURE(result)) {
+    ovr_Shutdown();
+    return;
+  }
+
+  state.lastButtonState = 0;
+  state.isInitialized = true;
+  state.isMirrored = true;
+  state.mirrorTexture = NULL;
+  int i;
+  for (i = 0; i < 2; i++) {
+    state.eyeTextures[i].size.w = 0;
+    state.eyeTextures[i].size.h = 0;
+    state.eyeTextures[i].fboId = 0;
+    state.eyeTextures[i].depthId = 0;
+    state.eyeTextures[i].chain = NULL;
+  }
+  state.clipNear = 0.1f;
+  state.clipFar = 30.f;
+
+  vec_init(&state.controllers);
+
+  for (i = 0; i < 2; i++) {
+    Controller *controller = lovrAlloc(sizeof(Controller), lovrControllerDestroy);
+    controller->id = ovrHand_Left+i;
+    vec_push(&state.controllers, controller);
+  }
+
+  lovrEventAddPump(lovrHeadsetPoll);
+  atexit(lovrHeadsetDestroy);
+}
+
+void lovrHeadsetDestroy() {
+  int i;
+  Controller *controller;
+  vec_foreach(&state.controllers, controller, i) {
+    lovrRelease(&controller->ref);
+  }
+
+  vec_deinit(&state.controllers);
+
+  if (state.mirrorTexture) {
+    ovr_DestroyMirrorTexture(state.session, state.mirrorTexture);
+    state.mirrorTexture = NULL;
+  }
+
+  for (i = 0; i < 2; i++) {
+    if (state.eyeTextures[i].chain) {
+      ovr_DestroyTextureSwapChain(state.session, state.eyeTextures[i].chain);
+      state.eyeTextures[i].chain = NULL;
+    }
+  }
+
+  ovr_Destroy(state.session);
+  ovr_Shutdown();
+
+  state.isInitialized = false;
+  state.frameIndex = 0;
+}
+
+static void checkInput(Controller *controller, int diff, int state, ovrButton button, ControllerButton target) {
+  if ((diff & button) > 0) {
+    Event e;
+    if ((state & button) > 0) {
+      e.type = EVENT_CONTROLLER_PRESSED;
+      e.data.controllerpressed.controller = controller;
+      e.data.controllerpressed.button = target;
+    }
+    else {
+      e.type = EVENT_CONTROLLER_RELEASED;
+      e.data.controllerreleased.controller = controller;
+      e.data.controllerreleased.button = target;
+    }
+    lovrEventPush(e);
+  }
+}
+
+void lovrHeadsetPoll() {
+  ovr_SetTrackingOriginType(state.session, ovrTrackingOrigin_FloorLevel);
+  state.ts = ovr_GetTrackingState(state.session, 0, true);
```

Comment:
Since we don't use the tracking state to emit any events, is it possible to be more lazy here with respect to the tracking state and fetch it in `lovrHeadsetGetPosition` etc.?

The event pump isn't really meant to be used for general updating (though it's currently the only place to do update-type stuff).

Reply:
I think you're only supposed to call GetTrackingState once per frame since it does some stuff with latency markers, and beyond that I have no idea how many times per frame a user might call the functions for TS, so it'd end up shuffling a bunch of data around needlessly.

Reply:
checked the docs again and I need to call GetTrackingState with the third argument true once per frame for accurate latency measurements. I was actually using it wrong here to begin with, however, the guide says I should be using predicted time instead of real (0). fixed in latest.